### DISCUSSION TITLE
LightMode Options

### DISCUSSION THREAD

**Discussion by brownjd** [aqualinkd/AqualinkD#177] [Q&A] [answered]
I am wondering about LightMode= options in aqualinkd.conf for incandescent lights. What are the various options and what do they do?

Does aqualinkd support incandescent light dimmers, like [this one](https://www.amazon.com/Jandy-AquaLink-Incandescent-Dimming-Module/dp/B002NJSLKM)?

**Comment by sfeakes** [comment]
Aqualinkd supports lights that work by sending on/off pulses. I’ve not seen that type of light controller before, but every light I’ve come across works with on/off pulses.  Aqualinkd can either tell the control panel to send the pulses (for control panels that support colored lights) or it can send the pulses itself so you can use colored lights on control panels that have no support.
I’ve never seen a dimmable single color light controller like the link.  But I assume it works by sending it a series of on/off pulses.  If that’s accurate than you might be able to simply change the aqualinkd config to the specific pulse timings of that light.    But it’s never need tested to my knowledge and I have no way to test it.

**Comment by brownjd** [comment]
The installation instructions indicate that you simply specify that the AUX (AUX2 in this case) is a dimmer, under System Setup -> Set Dimmers.

The dimmer does only support 4 brightness settings, 25%, 50%, 75%, 100%. The dimming slider in the iAqualink app moves in steps.

So, I’m not sure about sending pulses. I’d imagine that pulsing a relay would wear it out really fast.

Also, what do the LightMode settings do? I didn't see it in the documentation.

Thoughts?

**Comment by sfeakes** [comment]
The on off pulse is simply to set the mode, it’s not like PWM where it’s a constant on/off at high rate to limit current or voltage. So no it doesn’t wear out the relay.
The light mode and light pulse are explained in the wiki, go to the AqualinkD service configuration section.

**Reply by brownjd** [reply]
I haven't had any luck with setting the pulses. You'll see I posted my results with pulseview at the bottom. Any suggestions?

**Comment by sfeakes** [comment]
The below is from the manual, so yes it should work as it is controlled by pulse.  But I don’t know anyone that has actually used that specific board with Aqualinkd.
—————————-
Operation:The Controller signals the Light dimming Relay to turn 'fully on", "partially dimmed", or "completely off" by sending aspecific number of pulses in a set period of time to the Light Dimming Relay. The circuitry of the Light Dimming Relay(which requires 120 VAC to operate) then interprets the number and timing of pulses to determine whether the light shouldbe on, off, or dimmed.

**Reply by brownjd** [reply]
Yep, and I've measured them now, but I'm lost on how to apply the settings described in the wiki. See my comment below with the details of my results with PulseView.

**Comment by brownjd** [comment]
Thanks for checking - I will play with the light programming. For now, it seems to remember the dim setting, so I can set it through the iAqualink app then use aqualinkd to simply turn it on and off.

If you using aqualinkd to control, then you probably lose the functionality to control dim settings through iAqualink, correct?

Jason

**Comment by sfeakes** [comment]
No, you shouldn't loose any functionality by using AqualinkD to program the light. (iAqualink will still be able to program it)
Since your light is something I have not come across, you will need to make your own config for the UI, once you get the aqualinkd.conf sorted, you'll also need to modify config.js   Info is in the wiki.

**Reply by brownjd** [reply]
A couple of questions:

I have tried every combination of light_programming_mode, light_programming_initial_on, and light_programming_initial_off with no change in behavior - other than a lot of blinking at the same dimness. I know the settings are doing something because the timing changes, but all it does is cause the light to flash. I do have the button light mode set to 0. 1 and 2 didn't work.

Next, I did update config.js, but what I don't understand is what those labels correspond to. I get that they change the UI when you long press, but do they have to match some know label for the controller? Reason being, I notice that when I use iAqualink to set the dimness, aqualinkd will report a percentage in the status area in the web view. like "25% *~". 

Can I send you a serial debug where I run through the various dim settings in an effort to figure out how this thing works? It would be nice to have aqualinkd actually report the current brightness if the light is on. iAqualink IOS client does so.

**Comment by sfeakes** [comment]
You will need to read the light manual to get the specific of the pulse.
But first you want aqualinkd to program
button_XX_lightmode=0

then you will need to change the following to what the manual states.
light_programming_initial_on=
light_programming_initial_off=

start with
light_proframming_mode=0
Then once working you can play with that to make it faster.


The wiki has the description of the above, but will be no good if you don’t get the information from the light manual on how long the initial on and off needs to be to put the light into programming mode.


The files you mention will change how the UI displays the options.  So the first in the list will be 1 pulse, second 2 pulses and so on.   So if 1 pulse is 5% and 2 is 10% you will want a list like
[
“5 %”,
“10 %”,
“15 %”,
………
]

**Reply by brownjd** [reply]
> You will need to read the light manual to get the specific of the pulse. 

You quoted the manual above. I did not see any specifications for pulses in that same manual. Should I be looking elsewhere? I assumed I would need to debug the actual operation using iAqualink to see what the pulses are. It spews so much output that I could not make heads or tails of it.

>But first you want aqualinkd to program button_XX_lightmode=0
> 
> then you will need to change the following to what the manual states. light_programming_initial_on= 

The wiki indicates that light_programming_initial_on is used when switching to light_show, which in my case won't have any entries and won't be used. Since it is monochromatic, I'm assuming I'm only using entries on the left side of the config.js rendered box. So is this setting really needed?

When I use the jandy controller, the light comes on at the specified brightness. It does not come on, then click up or down in steps. It also does not turn off first or flicker.


> light_programming_initial_off=
> 


> start with light_proframming_mode=0 Then once working you can play with that to make it faster.

To be clear, I did start with programming_mode=0 and then played with various entries for initial on and initial off with my entries in config.js being: 25%, 50%, 75%, and 100%. Depending on the duration, it might flicker the light on and off for a second, but it never seemed to change the actual brightness.

> 
> The wiki has the description of the above, but will be no good if you don’t get the information from the light manual on how long the initial on and off needs to be to put the light into programming mode.

I wish the manual specified these. 

> 
> The files you mention will change how the UI displays the options. So the first in the list will be 1 pulse, second 2 pulses and so on. So if 1 pulse is 5% and 2 is 10% you will want a list like [ “5 %”, “10 %”, “15 %”, ……… ]

This is what I did. It only has 4 steps when on, 25, 50, 75, and 100.

**Reply by brownjd** [reply]
After multiple google searches, I have not located the timing for the dimmer. Probably for the very same reason you had to reverse engineer their protocol. They don't publish it.

Is there any way to get use the serial debug output to get the pulse timing and see what it is doing?

**Comment by sfeakes** [comment]
The serial debugger will only read RS485.  The control panel is pulsing the light relay directly, so their is no way the serial debugger will pick that up.  You might want to look at the timings of the color lights from Jandy.  I know pantair publishes the timings for color lights, and Jandy might do the same.  If so you could try those.

**Reply by john1158way** [reply]
You can purchase a cheap logic analyzer on Amazon for $10 and determine pulse timings   It will be a sigrok clone that runs pulse view free software.

**Reply by brownjd** [reply]
Thanks - do you have one you like?

**Reply by john1158way** [reply]
Comidox 1Set USB Logic Analyzer Device Set USB Cable 24MHz 8CH 24MHz 8 Channel UART IIC SPI Debug for Arduino ARM FPGA M100 $10.99 on Amazon

**Reply by brownjd** [reply]
Thanks!

**Reply by brownjd** [reply]
So I did this, see my answer at the bottom. Do you have any idea what settings to use in aqualinkd.conf based on the timings I observed?

**Comment by sfeakes** [comment]
When you change the value of the light from the control panel, do you see it flash?  Also you can play with it and kind-a guess simply by turning it on/off manually.   I had a look at parts / firmware / boards I have around and I can't even get anything to simulate / or even a control panel to think it's got anything like that light board.

I went back to the manual, and found this
https://www.manualslib.com/manual/805967/Jandy-Aqualink-Rs.html?page=9#manual
```
From the Controller - When the auxiliary, which has been assigned to light dimming is pressed or programmed on,the light will turn on to the brightness last set by pressing Back or Forward Buttons. Once on, the brightness maybe changed by pressing the Back or Forward Buttons. To turn it off, press the auxilliary button again.
From the Power Center override switch - Pressing the appropriate auxilliary button will turn the light on to thebrightness last set at the Controller or Spa Side Switch.
From the Spa Side Switch - If a button on the Spa Side Switch has been programmed to control a Light Dimmerauxilliary, the first press of this button will turn the light on to full brightness, each additional press will dim the lightin 25% increments until the light is off.
```

Do you have a standard keypad?  if so what happens when you turn it on, wait a few seconds and press the button again, does it dim, or turn off? (if no keypad, set the light in aqualinkD to standard switch and use that)

**Reply by brownjd** [reply]
Unfortunately, no keypad. All I have is the iAqualink app and the panel itself. I did see that in the manual, and indeed, that is how it works. Using the app, I can set the brightness directly. When toggled, the light will come back on the to last brightness setting. There is no forward or backward in the app, it is a slider with stops for 0, 25, 50, 75, 100.

As to your question about behavior, it will turn on or off. That means, it you flick the IOS app switch, it will turn on to the last brightness. If you click again, it will turn off. It will also report what brightness it is set to (I have seen this message displayed in the aqualinkd web interface, fyi). There is a > next to the on off switch while will bring up the slider to set the brightness. If you say go from 25% -> 75%, it will go directly to that brightness without flickering (mostly - sometimes I have seen a brief dimming or flickering - perhaps due to the timing).

**Comment by brownjd** [comment]
I hooked it up to pulseview to see what it does. As a refresher, the dimmer supports 25%, 50%, 75%, 100%.

From OFF:

25% - 2 pulses of 60ms per pulse, with a 8ms break between them
50% - 4 pulses of 60ms per pulse, with 8 ms break between them
75% - 6 pulses of 60ms per pulse, but the 8 ms break, but there is some variability - I'm guessing measurement error.
100% - 8 pulses of 60ms per pulse, but the breaks vary from 13ms (first), 3ms (second), 13ms (third), 13ms (fourth), 3ms (fifth), 4ms (sixth), 8.7ms (last break)
From 25% On -> 50%

4 pulses, with 3ms break, 13ms break, 7ms break
From 50% -> 75%

6 pulses
From 100% -> 25%

2 pulses
From 100% to OFF

1 pulse

So, with this information, what do I enter in aqualinkd.conf for these entries? Assuming a tolerance of 5-15ms between pulses, would the value of light mode be 0.010 (to pick 10 ms?). How to I account for the fact that my steps are in sets of 2? That is, 25 % is two pulses, 50% is 4 pulses, etc. Finally, the behavior is different than colored lights - that is, they don't need to come on first and then be programmed. The number of initial pulses turns them on to the desired brightness. One pulse turns them off.

I don't believe anything special has to happen to move up or. down in brightness. That is, I don't think it needs to turn off first or do a special sequence to turn on first. The brightness settings are in pairs and off is a single pulse.

button_XX_lightmode=0.010
light_programming_initial_on=?
light_programming_initial_off=?

**Reply by brownjd** [reply]
Any help on this?

**Comment by sfeakes** [comment]
Version 2.3.0c now supports dimmers.  Light mode 6.  So in aqualinkd config use something like the below.
```
button_07_label=Test Light
button_07_lightMode=6
```

**Reply by brownjd** [reply]
Thanks for this. 

Testing it with 2.3.2 yields the following results: Aqualink web UI correctly reports the ability to set brightness percentage, however clicking the button does not turn on the lights. The light will revert back to the off setting after a second or two (presumably after it polls and gets the current status of off). So, it appears the light mode is correct as far as aqualinkd is concerned, but it is not actually turning on the light. 

Are there additional prerequisites in aqualinkd.conf? Here's what I think are the relevant parts of the log:

Notice:  AqualinkD: Aqualink Daemon v2.3.2
Notice:  AqualinkD: Panel set to RS-4 Pool/Spa Only
Notice:  AqualinkD: Config log_level         = 6
Notice:  AqualinkD: Config device_id         = 0x0a
Notice:  AqualinkD: Config rssa_device_id    = 0x48
Notice:  AqualinkD: Config extra_device_id   = 0x31
Notice:  AqualinkD: Config extra_device_prog = YES
Notice:  AqualinkD: Config serial_port       = /dev/ttyUSB0
Notice:  AqualinkD: Config socket_port       = 80
Notice:  AqualinkD: Config web_directory     = /var/www/aqualinkd/
Notice:  AqualinkD: Config use_aux_labels    = NO
Notice:  AqualinkD: Config override frz prot = NO
Notice:  AqualinkD: Config mqtt_server       = localhost:1883
Notice:  AqualinkD: Config mqtt_dz_sub_topic = (null)
Notice:  AqualinkD: Config mqtt_dz_pub_topic = (null)
Notice:  AqualinkD: Config mqtt_aq_topic     = aqualinkd
Notice:  AqualinkD: Config mqtt_user         = (null)
Notice:  AqualinkD: Config mqtt_passwd       = (null)
Notice:  AqualinkD: Config mqtt_ID           = aqualinkd_b827eba3
Notice:  AqualinkD: Config force SWG         = NO
Notice:  AqualinkD: Config force PS setpoint = NO
Notice:  AqualinkD: Config force Freeze Prot = NO
Notice:  AqualinkD: Config deamonize         = NO
Notice:  AqualinkD: Config log_file          = (null)
Notice:  AqualinkD: Config enable scheduler  = YES
Notice:  AqualinkD: Config light_pgm_mode    = 0.00
Notice:  AqualinkD: Debug RS485 protocol     = NO
Notice:  AqualinkD: Debug RS485 protocol raw = NO
Notice:  AqualinkD: Display warnings in web  = YES
Notice:  AqualinkD: Keep panel time in sync  = YES
Notice:  AqualinkD: Read SWG direct          = NO
Notice:  AqualinkD: Read ePump direct        = YES
Notice:  AqualinkD: Read vsfPump direct      = NO
Notice:  AqualinkD: Serial FTDI low latency  = YES
Notice:  AqualinkD: Config BTN Filter_Pump   = label Filter Pump     | VSP ID 0x78 | PumpID 1 |
Notice:  AqualinkD: Config BTN Aux_1         = label Cleaner         | 
Notice:  AqualinkD: Config BTN Aux_2         = label Pool Light      | Light Progm | CTYPE 6  |
Notice:  AqualinkD: Config BTN Aux_3         = label Sheer Descent   | 
Notice:  AqualinkD: Config BTN Pool_Heater   = label Pool Heater     | 
Notice:  AqualinkD: Config BTN Spa_Heater    = label NONE            | 
Notice:  AqualinkD: Config BTN Solar_Heater  = label Solar Heater    | 
.....
Info:    iAQ Touch: Button 04|           Pool Light  | type=0x07 | state=0x00 | unknown=0x00 | keycode=0x15

Here is the log output when you click the light button in the web ui

Info:    AqualinkD: received 'ON' for 'Aux_2', turning 'ON'
Info:    JandyDvce: Jandy Pump Status message = WATTS 2222
Info:    RS Allbtn: RS Message :- 'Air Temp 100 F'
Info:    RS Allbtn: RS Message :- 'Water Temp 88 F'
Info:    NetService:MQTT: Published id=59: aqualinkd/Filter_Pump/Watts 2222
Info:    iAQ Touch: Page: HOME | 0x01
Info:    iAQ Touch: Status page 00| 88  
Info:    iAQ Touch: Status page 01| 100  
Info:    iAQ Touch: Status page 05| Air Temp
Info:    iAQ Touch: Page: HOME | 0x01
Info:    iAQ Touch: Status page 00| 88  
Info:    iAQ Touch: Status page 01| 100  
Info:    iAQ Touch: Status page 05| Air Temp
Info:    RS Allbtn: RS Message :- '08/05/23 Sat'
Info:    JandyDvce: Jandy Pump Status message = RPM 3450
Info:    RS Allbtn: RS Message :- '11:47 AM'
Info:    RS Allbtn: RS Message :- 'Air Temp 100 F'
Info:    RS Allbtn: RS Message :- 'Water Temp 88 F'
Info:    iAQ Touch: Page: HOME | 0x01
Info:    iAQ Touch: Status page 00| 88  
Info:    iAQ Touch: Status page 01| 100  
Info:    iAQ Touch: Status page 05| Air Temp
Info:    JandyDvce: Jandy Pump Status message = WATTS 2222

**Comment by brownjd** [comment]
Further testing yielded user error. I see there is a new option in the UI for light runtime that did not previously exist. If you turn on the lights, you have to move it off of the default 0 or turning them on will not turn them on. It seems that 0 should correspond to 'no timer,' that is, always on. If you just click the tile (rather than long press), you don't know that this is an option and the tile will not work.

Perhaps the light timer and the default duration could be configured in aqualinkd.conf?

One other item of note, if you pick a dim setting, say 50% and 360 minutes, they will turn on at the appraise level, but you see the following error message in UI:

Error: panl prog: light programmings didn't receive off message 

Then, how do you configure the homebridge plugin to see the light switch as a dimmer?

**Comment by sfeakes** [comment]
Seems like there may be an opportunity to make the UI more user friendly, I’ll take a look when I get the opportunity. Can’t for the next few weeks due to travel.   Seems like what you are suggesting, is how I thought it was working, so there may be a bug in the UI.

There is no option to program / set these in HomeKit.  It only supports on/off for lights and switches, no timers, no light modes and no dimmers.

**Comment by brownjd** [comment]
I think HomeKit does support dimmers, because I use them for other switches. I think you just have to expose a new type from the api so that your plugin knows it is dealing with a lightbulb, then you can use: https://developers.homebridge.io/#/service/Lightbulb

Even if on/off/dim were exposed through the homekit plugin, it could just use the default 0/no timer setting. In other words, just don’t expose the timer, but allow dimming.

Jason

**Comment by sfeakes** [comment]
HomeKit does support dimmers, but there are two problems here. I’m not sure you can set HomeKit to a 0/25/50/75/100 only option, you can set increments in things like thermostats, but not sure on dimmers, and having an increment or 25. Then I’ll also need to distinguish a different light type between regular light and dimmable light. Also it’s a lot more involved in the homebridge-AqualinkD code than simply exposing the api.   It can be done, but will take some time.
Timer would never be supported unless I change the type to faucet, I believe that’s the only type in HomeKit that supports a timer.

edit.  I see my mistake.  
When I said “ There is no option to program / set these in HomeKit. It only supports on/off for lights and switches, no timers, no light modes and no dimmers.” in my previous messages.   I was referring to AqualinkD/HomeKit-AqualinkD when I said “it”.

**Comment by brownjd** [comment]
The API would know the light mode (1-6), so it could reveal the light type: dimmable or not to the plugin. Even if homekit or homebridge did not support discrete brightness steps, it could snap-to any of those. Meaning, if I set the slider to 33%, home bridge would pass along 33% to the API, but the api could just set it to the next closest step. When the accessory polls for the current setting, it would correctly reflect 25 or 50, whatever was chosen. So I might pick 33 in the UI, but several seconds later it will correctly reflect the correct value.

**Reply by brownjd** [reply]
Any thoughts on this approach?

**Reply by brownjd** [reply]
There have been several revisions since this thread ended. Were any updates made to support the non-color dimmers?

**Comment by sfeakes** [comment, marked as answer]
Dimmable lights were added in the following versions.
Homekit-aqualinkD v0.3.5 (2024-10-12)
AqualinkD v2.5.0

**Reply by brownjd** [reply]
Yes they were. Thank you!

**Reply by brownjd** [reply]
How do you expose the correct control in HomeKit? It still displays as a on/off switch in HomeKit when the aqualinkd web interface allows brightness settings.

For those following along at home, Lightmode 10 is what you want.
